### PR TITLE
Refactor providers

### PR DESCRIPTION
**Pull request by sscargal**
To make things easier to maintain I've refactored the existing providers in to their own files under src/pvdr.  No major changes were made.  The uprobe and kprobe providers need more work to produce cleaner maintainable code.  eg kprobe_setattach() has uprobe checks hard coded.

```
$ ls -1
kprobe.c
kprobe.h
profile.c
pvdr.c
tracepoint.c
uprobe.c
```

Added -P/--providers flag which shows available (registered) providers.

```
Options:
  -A                  ASCII output only, no Unicode.
  -c <script_string>  Execute script literate.
  -d                  Enable debug output.
  -D                  Dump generated BPF and exit.
  -h                  Print usage message and exit.
  -P                  Print available Providers.
  -t <timeout>        Terminate trace after <timeout> seconds.
  -v                  Print version information.
```

The -P/--providers output now displays the name and a short description

```
$ ./ply -P
            PROVIDER DESCRIPTION
              kprobe Kernel function entry provider.
           kretprobe Kernel function return provider.
             profile Profiles with time-based intervals.
              uprobe Userland function entry provider.
           uretprobe Userland function return provider.
```

### CODE REVIEW

**Comment by alan-maguire** [comment]
The change is a great idea; it looks much cleaner. Thanks for doing this! One question, not directly related to the refactoring; you've added a bunch of extensions to the man/.gitignore file; don't they belong in a .gitignore the toplevel of the repo? Also given the number of providers in place, feels like some unit tests for ply would be a good idea in the future.

**Comment by sscargal** [comment]
Ouch!  Good catch on the .gitignore issue.  I'll address it with a new commit to the branch.

I agree on unit testing each provider.

**Comment by sscargal** [comment]
I've fixed the .gitingore issue.  Thanks again!

File: src/pvdr/kprobe.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+	struct {
+		int cap, len;
+		int *fds;
+	} efds;
+} kprobe_t;
+
+extern const module_t *kprobe_modules[];
+extern const module_t *kretprobe_modules[];
+
```
Comment (wkz) [inline src/pvdr/kprobe.h]: Many of the following functions are not used by other providers so don't expose them in the header.

If there are ordering issues, please put the required prototypes in `kprobe.c`.

File: src/pvdr/profile.c [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        .setup = profile_setup,
+        .teardown = profile_teardown,
+};
+
+/* Register Provider */
+__attribute__((constructor))
+static void profile_pvdr_register(void)
+{
+	pvdr_register(&profile_pvdr);
+}
```
Comment (wkz) [inline src/pvdr/profile.c]: Newline please.

File: src/pvdr/uprobe.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	.teardown = kprobe_teardown,
+};
+
+/* Register Provider */
+__attribute__((constructor))
+static void uprobe_pvdr_register(void)
+{
+	pvdr_register(   &uprobe_pvdr);
+	pvdr_register(&uretprobe_pvdr);
+}
```
Comment (wkz) [inline src/pvdr/uprobe.c]: Newline please.

**Comment by wkz** [comment]
Thanks for this contribution!

I have a few comments, as seen in the review, but first I would just like to give you a heads up that I'm working on a new version of ply. It's a from-scratch rewrite of the compiler to fix some of the most glaring holes in the first version, there's now a proper (eventually) C compatible type system, and an intermediate representation making instruction selection much easier. I will still accept PRs to the current master. Just beware that this change is coming.

**Pull request closed without merging**